File: go-scanner/token/token.go
//go:generate go tool stringer -type=Type
package token

import "fmt"

type Type uint

const (
	// Single-character tokens.
	LEFT_PAREN Type = iota
	RIGHT_PAREN
	MINUS
	PLUS
	SLASH
	STAR

	// One or two character tokens.
	BANG
	BANG_EQUAL
	EQUAL
	EQUAL_EQUAL
	GREATER
	GREATER_EQUAL
	LESS
	LESS_EQUAL

	// Literals.
	IDENTIFIER
	NUMBER

	EOF
)

type Token struct {
	Type    Type
	Lexeme  string
	Literal any

	Line   int
	Column int
}

func (t Token) String() string {
	return fmt.Sprintf("[%d:%d] %s %s %v", t.Line, t.Column, t.Type, t.Lexeme, t.Literal)
}

File: syntax-highlighter/src/token.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenType {
    LeftParen,
    RightParen,
    Minus,
    Plus,
    Slash,
    Star,

    // One or two character tokens
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals
    Identifier,
    Number,

    Eof,
}

pub struct Token {
    pub token_type: TokenType,
    pub lexeme: String,
    pub literal: Option<i32>,
    pub line: usize,
    pub column: usize,
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}:{} {:?} '{}' {}",
            self.line,
            self.column,
            self.token_type,
            self.lexeme,
            self.literal.map_or(String::new(), |v| v.to_string())
        )
    }
}
